File: packages/ai/ai_core/aircanvas/shape_recognizer.py
import numpy as np
from typing import List, Dict, Any, Tuple

# ==============================================================================
# Ramer-Douglas-Peucker (RDP) Trajectory Simplification
# ==============================================================================

def perpendicular_distance(pt: np.ndarray, start: np.ndarray, end: np.ndarray) -> float:
  """
  Calculates the perpendicular distance from pt to the line segment start -> end.
  """
  if np.allclose(start, end):
    return float(np.linalg.norm(pt - start))

  line_vec = end - start
  pt_vec = pt - start
  line_len = np.linalg.norm(line_vec)
  line_unit = line_vec / line_len

  proj_len = np.dot(pt_vec, line_unit)
  perp_vec = pt_vec - proj_len * line_unit
  return float(np.linalg.norm(perp_vec))


def rdp_simplify(points: np.ndarray, epsilon: float) -> np.ndarray:
  """
  Simplifies a curve of points using the Ramer-Douglas-Peucker algorithm.
  """
  if len(points) < 3:
    return points

  start = points[0]
  end = points[-1]

  dmax = 0.0
  index = 0

  for i in range(1, len(points) - 1):
    d = perpendicular_distance(points[i], start, end)
    if d > dmax:
      index = i
      dmax = d

  if dmax > epsilon:
    results1 = rdp_simplify(points[:index + 1], epsilon)
    results2 = rdp_simplify(points[index:], epsilon)
    return np.vstack((results1[:-1], results2))
  else:
    return np.array([start, end])


# ==============================================================================
# Shape Classifier Logic
# ==============================================================================

def recognize_shape(stroke_points: List[Dict[str, Any]]) -> Tuple[str, float, Dict[str, Any]]:
  """
  Recognizes geometric shapes (LINE, CIRCLE, RECTANGLE, TRIANGLE, ARROW)
  from a list of coordinate points.
  Returns: (shape_type, confidence, bounding_box)
  """
  N = len(stroke_points)
  if N < 5:
    return "UNKNOWN", 0.0, {"minX": 0, "minY": 0, "maxX": 0, "maxY": 0}

  xs = np.array([p["x"] for p in stroke_points])
  ys = np.array([p["y"] for p in stroke_points])

  min_x, max_x = float(np.min(xs)), float(np.max(xs))
  min_y, max_y = float(np.min(ys)), float(np.max(ys))
  width = max_x - min_x
  height = max_y - min_y
  max_dim = max(width, height)

  bbox = {
    "minX": min_x,
    "minY": min_y,
    "maxX": max_x,
    "maxY": max_y
  }

  if max_dim < 1e-5:
    return "UNKNOWN", 0.0, bbox

  # 1. Scale and center coordinates to fit [-1.0, 1.0] for scale-invariant checks
  centroid_x = np.mean(xs)
  centroid_y = np.mean(ys)
  scale_factor = 2.0 / max_dim

  scaled_x = (xs - centroid_x) * scale_factor
  scaled_y = (ys - centroid_y) * scale_factor
  scaled_pts = np.column_stack((scaled_x, scaled_y))

  # 2. Check if path is closed
  start_pt = scaled_pts[0]
  end_pt = scaled_pts[-1]
  dist_start_end = np.linalg.norm(start_pt - end_pt)
  is_closed = dist_start_end < 0.40  # Path closes near start point

  # Calculate path length
  diffs = np.diff(scaled_pts, axis=0)
  path_len = np.sum(np.sqrt(np.sum(diffs * diffs, axis=1)))

  # 3. Circle detector (check radius uniformity from centroid)
  radii = np.sqrt(scaled_x*scaled_x + scaled_y*scaled_y)
  r_mean = np.mean(radii)
  r_std = np.std(radii)
  r_cv = r_std / r_mean if r_mean > 0 else 99.0

  if is_closed and r_cv < 0.12:
    # High radius uniformity indicates circle
    confidence = float(max(0.0, 1.0 - r_cv * 2.0))
    return "CIRCLE", confidence, bbox

  # 4. RDP simplification for lines, triangles, and rectangles
  epsilon = 0.08
  simplified = rdp_simplify(scaled_pts, epsilon)
  V = len(simplified)

  # 5. Line detector
  if not is_closed:
    # Check max perpendicular deviation from straight line connecting start and end
    deviations = [perpendicular_distance(pt, start_pt, end_pt) for pt in scaled_pts]
    max_dev = max(deviations)
    if max_dev < 0.15:
      # Straight line
      confidence = float(max(0.0, 1.0 - max_dev * 2.5))
      return "LINE", confidence, bbox

  # 6. Triangle detector
  if is_closed and (V == 4 or V == 3):
    # Simplified vertices are 3 corners (plus close point returning V=4)
    return "TRIANGLE", 0.85, bbox

  # 7. Rectangle detector
  if is_closed and (V == 5 or V == 6):
    # Simplified vertices are 4 corners (plus close point returning V=5 or 6)
    # Check angles at vertices to confirm approximately right angles
    return "RECTANGLE", 0.85, bbox

  # 8. Arrow detector (shaft + head)
  if not is_closed and N > 10:
    # An arrow has a long line (shaft) followed by a sharp turnaround (cusp) at one end
    # We find if there is a sharp corner (>135 degrees) in the last 30% of the path
    angles = []
    for i in range(int(N * 0.7), N - 2):
      v1 = scaled_pts[i] - scaled_pts[i - 1]
      v2 = scaled_pts[i + 1] - scaled_pts[i]
      len1 = np.linalg.norm(v1)
      len2 = np.linalg.norm(v2)
      if len1 > 1e-4 and len2 > 1e-4:
        cos_theta = np.dot(v1, v2) / (len1 * len2)
        angle = np.arccos(np.clip(cos_theta, -1.0, 1.0)) * 180.0 / np.pi
        angles.append(angle)
    
    # If there is a sharp turnaround (large angle change)
    if angles and max(angles) > 110.0:
      return "ARROW", 0.80, bbox

  # 9. Fallbacks for closed shapes that might simplify slightly differently
  if is_closed:
    if V == 7 or V == 8:
      return "RECTANGLE", 0.65, bbox

  return "UNKNOWN", 0.0, bbox

File: apps/frontend/src/utils/shapeRecognizer.ts
/**
 * AIRSPACE — Geometric Shape Recognizer
 * Evaluates hand-drawn stroke coordinate trajectories to classify
 * and snap rough drawings into clean geometric shapes:
 * - Circle
 * - Rectangle
 * - Triangle
 * - Line
 * - Arrow
 */

export interface Point {
  x: number;
  y: number;
  t?: number;
}

export interface BoundingBox {
  minX: number;
  minY: number;
  maxX: number;
  maxY: number;
  width: number;
  height: number;
  centerX: number;
  centerY: number;
}

export type ShapeType = "CIRCLE" | "RECTANGLE" | "TRIANGLE" | "LINE" | "ARROW" | "UNKNOWN";

export interface RecognizedShapeResult {
  type: ShapeType;
  confidence: number;
  bbox: BoundingBox;
  cleanPoints?: Point[];
  details?: any;
}

function perpendicularDistance(pt: Point, lineStart: Point, lineEnd: Point): number {
  const dx = lineEnd.x - lineStart.x;
  const dy = lineEnd.y - lineStart.y;
  const lineLenSq = dx * dx + dy * dy;

  if (lineLenSq < 1e-6) {
    const ddx = pt.x - lineStart.x;
    const ddy = pt.y - lineStart.y;
    return Math.sqrt(ddx * ddx + ddy * ddy);
  }

  const t = Math.max(0, Math.min(1, ((pt.x - lineStart.x) * dx + (pt.y - lineStart.y) * dy) / lineLenSq));
  const projX = lineStart.x + t * dx;
  const projY = lineStart.y + t * dy;

  const distDx = pt.x - projX;
  const distDy = pt.y - projY;
  return Math.sqrt(distDx * distDx + distDy * distDy);
}

export function rdpSimplify(points: Point[], epsilon: number): Point[] {
  if (points.length < 3) return points;

  let dmax = 0;
  let index = 0;
  const start = points[0];
  const end = points[points.length - 1];

  for (let i = 1; i < points.length - 1; i++) {
    const d = perpendicularDistance(points[i], start, end);
    if (d > dmax) {
      index = i;
      dmax = d;
    }
  }

  if (dmax > epsilon) {
    const left = rdpSimplify(points.slice(0, index + 1), epsilon);
    const right = rdpSimplify(points.slice(index), epsilon);
    return [...left.slice(0, -1), ...right];
  } else {
    return [start, end];
  }
}

export function calculateBoundingBox(points: Point[]): BoundingBox {
  let minX = Infinity;
  let maxX = -Infinity;
  let minY = Infinity;
  let maxY = -Infinity;

  for (const p of points) {
    if (p.x < minX) minX = p.x;
    if (p.x > maxX) maxX = p.x;
    if (p.y < minY) minY = p.y;
    if (p.y > maxY) maxY = p.y;
  }

  const width = maxX - minX;
  const height = maxY - minY;
  const centerX = minX + width / 2;
  const centerY = minY + height / 2;

  return { minX, maxX, minY, maxY, width, height, centerX, centerY };
}

export function recognizeShape(stroke: Point[]): RecognizedShapeResult {
  const N = stroke.length;
  if (N < 4) {
    return {
      type: "UNKNOWN",
      confidence: 0,
      bbox: calculateBoundingBox(stroke)
    };
  }

  const bbox = calculateBoundingBox(stroke);
  const maxDim = Math.max(bbox.width, bbox.height);

  if (maxDim < 20) {
    return { type: "UNKNOWN", confidence: 0, bbox };
  }

  const startPt = stroke[0];
  const endPt = stroke[N - 1];
  const dxStartEnd = endPt.x - startPt.x;
  const dyStartEnd = endPt.y - startPt.y;
  const distStartEnd = Math.sqrt(dxStartEnd * dxStartEnd + dyStartEnd * dyStartEnd);
  const isClosed = distStartEnd < maxDim * 0.35;

  // 1. Check for Line (open path with minimal perpendicular deviation)
  if (!isClosed) {
    let maxDev = 0;
    for (const p of stroke) {
      const dev = perpendicularDistance(p, startPt, endPt);
      if (dev > maxDev) maxDev = dev;
    }

    // Line check
    if (maxDev < maxDim * 0.14) {
      const cleanPoints: Point[] = [
        { x: startPt.x, y: startPt.y },
        { x: endPt.x, y: endPt.y }
      ];
      const confidence = Math.max(0.7, 1 - maxDev / (maxDim * 0.14));
      return { type: "LINE", confidence, bbox, cleanPoints };
    }

    // 2. Check for Arrow (straight shaft with a sharp turnaround or arrowhead near end)
    if (N > 12) {
      const shaftEndIdx = Math.floor(N * 0.7);
      const shaftStart = startPt;
      const shaftEnd = stroke[shaftEndIdx];

      let shaftDev = 0;
      for (let i = 0; i <= shaftEndIdx; i++) {
        const dev = perpendicularDistance(stroke[i], shaftStart, shaftEnd);
        if (dev > shaftDev) shaftDev = dev;
      }

      const shaftLen = Math.sqrt(
        (shaftEnd.x - shaftStart.x) ** 2 + (shaftEnd.y - shaftStart.y) ** 2
      );

      if (shaftDev < shaftLen * 0.25 && shaftLen > 30) {
        const angle = Math.atan2(shaftEnd.y - shaftStart.y, shaftEnd.x - shaftStart.x);
        const headLength = Math.min(25, shaftLen * 0.3);
        const headAngle = Math.PI / 6;

        const wing1: Point = {
          x: shaftEnd.x - headLength * Math.cos(angle - headAngle),
          y: shaftEnd.y - headLength * Math.sin(angle - headAngle)
        };
        const wing2: Point = {
          x: shaftEnd.x - headLength * Math.cos(angle + headAngle),
          y: shaftEnd.y - headLength * Math.sin(angle + headAngle)
        };

        const cleanPoints: Point[] = [
          shaftStart,
          shaftEnd,
          wing1,
          shaftEnd,
          wing2
        ];
        return { type: "ARROW", confidence: 0.85, bbox, cleanPoints };
      }
    }
  }

  // 3. Polygon check (RDP simplification to find sharp corners)
  const epsilon = maxDim * 0.08;
  const simplified = rdpSimplify(stroke, epsilon);
  const V = simplified.length;

  // Triangle: 3 corners + closed endpoint = 4 vertices
  if (isClosed && (V === 4 || V === 3)) {
    const pts = simplified.slice(0, 3);
    if (pts.length === 3) {
      const cleanPoints: Point[] = [...pts, pts[0]];
      return { type: "TRIANGLE", confidence: 0.88, bbox, cleanPoints };
    }
  }

  // Rectangle / Box: 4 corners + closed endpoint = 5 vertices
  if (isClosed && (V === 5 || V === 6)) {
    const cleanPoints: Point[] = [
      { x: bbox.minX, y: bbox.minY },
      { x: bbox.maxX, y: bbox.minY },
      { x: bbox.maxX, y: bbox.maxY },
      { x: bbox.minX, y: bbox.maxY },
      { x: bbox.minX, y: bbox.minY }
    ];
    return { type: "RECTANGLE", confidence: 0.88, bbox, cleanPoints };
  }

  // 4. Check for Circle (closed curved path with uniform radial distance from centroid and smooth contour)
  if (isClosed && N >= 10) {
    const cx = bbox.centerX;
    const cy = bbox.centerY;
    let sumR = 0;
    const radii: number[] = [];

    for (const p of stroke) {
      const r = Math.sqrt((p.x - cx) ** 2 + (p.y - cy) ** 2);
      radii.push(r);
      sumR += r;
    }

    const meanR = sumR / N;
    let variance = 0;
    for (const r of radii) {
      variance += (r - meanR) ** 2;
    }
    const stdR = Math.sqrt(variance / N);
    const cv = stdR / (meanR || 1);

    if (cv < 0.22 && meanR > 15) {
      const cleanPoints: Point[] = [];
      const steps = 40;
      for (let i = 0; i <= steps; i++) {
        const theta = (i / steps) * Math.PI * 2;
        cleanPoints.push({
          x: cx + meanR * Math.cos(theta),
          y: cy + meanR * Math.sin(theta)
        });
      }
      return {
        type: "CIRCLE",
        confidence: Math.max(0.75, 1 - cv * 2.5),
        bbox,
        cleanPoints,
        details: { radius: meanR, cx, cy }
      };
    }
  }

  // Fallback for rough rectangles with extra jitter vertices
  if (isClosed && V >= 7 && V <= 8) {
    const cleanPoints: Point[] = [
      { x: bbox.minX, y: bbox.minY },
      { x: bbox.maxX, y: bbox.minY },
      { x: bbox.maxX, y: bbox.maxY },
      { x: bbox.minX, y: bbox.maxY },
      { x: bbox.minX, y: bbox.minY }
    ];
    return { type: "RECTANGLE", confidence: 0.70, bbox, cleanPoints };
  }

  return { type: "UNKNOWN", confidence: 0, bbox };
}
